packages/utilities/src/asAsync.tsx

```diff
@@ -0,0 +1,58 @@
+
+import * as React from 'react';
+
+interface IAsyncState<TProps> {
+  component?: React.ComponentType<TProps> | undefined;
+}
+
+export interface IModuleWithDefault<T> {
+  default: T;
+}
+
+/**
+ * Produces a component which internally resolves asynchronously before first mount.
+ * The component passes all props through to the loaded component.
+ */
+export function asAsync<TProps>(
+  load: () => Promise<{ default: React.ComponentType<TProps>; }>
+): React.ComponentType<TProps>;
+export function asAsync<TProps>(
+  load: () => Promise<React.ComponentType<TProps>>
+): React.ComponentType<TProps>;
+export function asAsync<TProps>(
+  load: () => Promise<React.ComponentType<TProps> | IModuleWithDefault<React.ComponentType<TProps>>>
+): React.ComponentType<TProps> {
+  class Async extends React.Component<TProps, IAsyncState<TProps>> {
```

Comment:
couple downsides with this solution;

1. There isn't a way to render a placeholder during load. This can create janky looking scenarios.
2. If the import has the component as a named export, there isn't a way to provide the named export.

I think these may be difficult to solve using a HOC. Why not a component itself?

```tsx
<Async resolve={ () => import('thing') } name='NamedThing'>
  <Spinner/>
</Async>

or if you just want a simple async wrapper::

```tsx
export const AsyncTooltip = (props) => <Async resolve={ () => import('tooltip') } props={props} />;
```

Reply:
1. Placeholder functionality may be added easily, and later if we like.
2. If you have a named export, you do this:
```tsx
const AsyncDetailsList = asAsync(async () => (await import('./DetailsList')).DetailsList);
```
The whole purpose of using the HoC model is that this control appears passthrough to consumers, and so you may apply additional decorators with impunity.